### ISSUE TITLE
Preprocessor: unable to traverse "up"

### ISSUE THREAD

**Issue by vgaming** [wesnoth/wesnoth#2180]
Preprocessor cannot traverse "up" with the `..` notation. It becomes impossible to use relative paths AND directories.

For example, consider the following directory layout:

```
.
├── a
│   └── a.cfg
├── b
│   └── b.cfg
├── _info.cfg
└── _main.cfg
```

Now, if you're inside `a.cfg`, it's currently impossible to reference `b.cfg` by a relative path.

Please add the `..` notation? I understand the security concern of course. But then again, we might just check that resulting address lives inside `~add-ons`. Therefore the scope of files to be accessed is the same as with the `~add-ons` notation.

**Comment by Pentarctagon** [comment]
See #1033

**Comment by vgaming** [comment]
@Pentarctagon interesting. Well, the linked PR is not accepted yet, and it's rather big.

Maybe a smaller one can be made such as those proposed in comments? Like just resolving the full path and checking whether it's inside ~add-ons?

**Comment by GregoryLundberg** [comment]
The problem with chasing `../` is it's rife with security issues which have to be handled carefully. That is why the PR was not accepted.

We're pushing to 1.14 and, while we're not, strictly speaking, in a feature freeze, I'd suggest that something this major should wait until after 1.14 ships.

I've been through this before with other software. If you want it fast, you need to reserve a CVE or four to handle the mistakes.

So, patience. There's an open PR on it and it's not going to be closed .. just delayed.

**Comment by vgaming** [comment]
@GregoryLundberg Argee, these kind of changes should be done with extreme caution. Clearly defined boundaries help a lot here though ("only allow accessing files inside ~add-ons dir"). Anyway, understood.

**Comment by CelticMinstrel** [comment]
@vgaming My intention was to push #1033 (most likely with much further review) _after_ 1.14 has been released. So it's not going to be in 1.14, but it will be happening eventually.

**Comment by vgaming** [comment]
@CelticMinstrel yeah, I understood. Gregory said the same thing basically (feature freeze). No problem.

**Comment by Wedge009** [comment]
1.14 was long since released, now 1.16 is imminent...
